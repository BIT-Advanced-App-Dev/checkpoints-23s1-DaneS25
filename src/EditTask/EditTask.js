/* 
Edit Task component.
Click edit to edit an existing task in the list.
*/

import Modal from "../Modal/Modal"
import {useState} from 'react'
import './editTask.css'
import { doc, updateDoc } from "firebase/firestore"; // Import updateDoc function from Firestore library
import {db} from '../firebase'

function EditTask({open, onClose, toEditTitle, toEditDescription, id, userUid}) {

  // Set initial state variables 'toEdit'
  const [title, setTitle] = useState(toEditTitle)
  const [description, setDescription] = useState(toEditDescription)

  // Function to update firestore
  const handleUpdate = async (e) => {
    e.preventDefault()
    const taskDocRef = doc(db, `users/${userUid}/tasks`, id); // Get a reference to the document to update using the ID passed as a prop
    try{
      await updateDoc(taskDocRef, { // Update the document with the new values for title and description
        title: title,
        description: description
      })
      onClose()
    } catch (err) {
      alert(err)
    }
    
  }

  return (
    <Modal modalLabel='Edit Task' onClose={onClose} open={open}>
      <form onSubmit={handleUpdate} className='editTask' name='editTask'>
        <input type='text' name='title' onChange={(e) => setTitle(e.target.value.toUpperCase())} value={title}/>
        <textarea onChange={(e) => setDescription(e.target.value)} value={description}></textarea>
        <button type='submit' name='editTask'>Edit</button> {/* Button that will trigger the handleUpdate function when clicked */}
      </form> 
    </Modal>
  )
}

export default EditTask