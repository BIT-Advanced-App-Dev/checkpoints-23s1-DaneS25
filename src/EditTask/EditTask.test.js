import React from 'react';
import renderer from 'react-test-renderer';
import { act } from 'react-dom/test-utils';
import EditTask from './EditTask';

it('renders EditTask component correctly', () => {
  const onClose = jest.fn(); // Create a mock function for the onClose prop
  const toEditTitle = 'Test Title';
  const toEditDescription = 'Test Description';
  const id = 'test-id';
  const component = renderer.create(<EditTask onClose={onClose} open={true} toEditTitle={toEditTitle} toEditDescription={toEditDescription} id={id} />); // Create the component
  let tree = component.toJSON(); // Get the initial snapshot

  // Simulate user input
  act(() => {
    const titleInput = component.root.findByProps({ name: 'title' }); // Find the title input element
    titleInput.props.onChange({ target: { value: 'New Test Title' } }); // Update the title state variable
    const descriptionInput = component.root.findByType('textarea'); // Find the description textarea element
    descriptionInput.props.onChange({ target: { value: 'New Test Description' } }); // Update the description state variable
  });

  // Get a new snapshot after the state variables have been updated
  tree = component.toJSON();
  expect(tree).toMatchSnapshot();

  // Simulate form submission
  act(() => {
    const form = component.root.findByProps({ name: 'editTask' }); // Find the form element
    form.props.onSubmit({ preventDefault: jest.fn() }); // Submit the form with a mock event object
  });
});
