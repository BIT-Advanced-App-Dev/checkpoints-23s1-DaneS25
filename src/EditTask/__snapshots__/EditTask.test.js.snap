// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders EditTask component correctly 1`] = `
<div
  className="modalContainer"
  onClick={[Function]}
>
  <div
    className="modal undefined"
  >
    <div
      className="modal__head"
    >
      <h2>
        Edit Task
      </h2>
      <span
        className="modal__close"
        onClick={[MockFunction]}
      >
        x
      </span>
    </div>
    <form
      className="editTask"
      name="editTask"
      onSubmit={[Function]}
    >
      <input
        name="title"
        onChange={[Function]}
        type="text"
        value="NEW TEST TITLE"
      />
      <textarea
        onChange={[Function]}
        value="New Test Description"
      />
      <button
        name="editTask"
        type="submit"
      >
        Edit
      </button>
       
    </form>
  </div>
</div>
`;
